File: app/src/test/java/com/test/casenio/wifi/WifiConnectivityReceiverTest.java
package com.test.casenio.wifi;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class WifiConnectivityReceiverTest {

    @Mock
    private WifiHelper mHelperMock;

    @Mock
    private ConnectivityListener mListenerMock;

    private WifiConnectivityReceiver mReceiver;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        mReceiver = new WifiConnectivityReceiver(mHelperMock, mListenerMock);
    }

    @Test
    public void onReceive_connectedToWifi() {
        when(mHelperMock.isConnected()).thenReturn(true);
        when(mHelperMock.isConnectedToWifi()).thenReturn(true);

        mReceiver.onReceive(null, null);

        verify(mListenerMock).successfulConnected();
        verify(mHelperMock, never()).disconnect();
    }

    @Test
    public void onReceive_connectedToDifferentWifi() {
        when(mHelperMock.isConnected()).thenReturn(true);
        when(mHelperMock.isConnectedToWifi()).thenReturn(false);

        mReceiver.onReceive(null, null);

        verify(mListenerMock, never()).successfulConnected();
        verify(mHelperMock).disconnect();
    }

    @Test
    public void onReceive_notConnected() {
        when(mHelperMock.isConnected()).thenReturn(false);
        when(mHelperMock.isConnectedToWifi()).thenReturn(false);

        mReceiver.onReceive(null, null);

        verify(mListenerMock, never()).successfulConnected();
        verify(mHelperMock, never()).disconnect();
    }
}

File: app/src/test/java/com/test/casenio/wifi/WifiConnectivityReceiverTest.kt
package com.test.casenio.wifi

import android.content.Context
import android.content.Intent
import org.junit.Before
import org.junit.Test

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations

class WifiConnectivityReceiverTest {

    @Mock
    private lateinit var helperMock: WifiHelper
    @Mock
    private lateinit var listenerMock: ConnectivityListener
    @Mock
    private lateinit var context: Context
    @Mock
    private lateinit var intent: Intent
    private lateinit var receiver: WifiConnectivityReceiver

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        receiver = WifiConnectivityReceiver(helperMock, listenerMock)
    }

    @Test
    fun onReceive_connectedToWifi() {
        `when`(helperMock.connected).thenReturn(true)
        `when`(helperMock.connectedToWifi).thenReturn(true)

        receiver.onReceive(context, intent)

        verify(listenerMock).successfulConnected()
        verify(helperMock, never()).disconnect()
    }

    @Test
    fun onReceive_connectedToDifferentWifi() {
        `when`(helperMock.connected).thenReturn(true)
        `when`(helperMock.connectedToWifi).thenReturn(false)

        receiver.onReceive(context, intent)

        verify(listenerMock, never()).successfulConnected()
        verify(helperMock).disconnect()
    }

    @Test
    fun onReceive_notConnected() {
        `when`(helperMock.connected).thenReturn(false)
        `when`(helperMock.connectedToWifi).thenReturn(false)

        receiver.onReceive(context, intent)

        verify(listenerMock, never()).successfulConnected()
        verify(helperMock, never()).disconnect()
    }
}